tools/pybind11Tools.cmake

```diff
@@ -18,27 +18,116 @@ find_package(PythonLibsNew ${PYBIND11_PYTHON_VERSION} REQUIRED)
+
+## Don't do anything if CMAKE_CXX_STANDARD is set
+if(NOT CMAKE_CXX_STANDARD)
+  # While complile features were introduced in 3.1, only 3.8+ have meta-features
+  # (And C++17 seems to be mostly supported through the meta-feature)
+
+  if(CMAKE_VERSION VERSION_LESS 3.8 AND NOT PYBIND11_CXX_FEATURES)
+    if(NOT PYBIND_CPP_STANDARD)
+      # Only try to get the standard manually if CMake doesn't support compiler features
+      if(MSVC)
+        set(PYBIND11_CPP_STANDARD "/std:c++14")
+        message(STATUS "pybind11 selected C++14 flag, MSVC")
+      else()
+        check_cxx_compiler_flag("-std=c++17" HAS_CPP17_FLAG)
+        check_cxx_compiler_flag("-std=c++14" HAS_CPP14_FLAG)
+        check_cxx_compiler_flag("-std=c++11" HAS_CPP11_FLAG)
+
+        if(HAS_CPP17_FLAG)
+          set(PYBIND11_CPP_STANDARD "-std=c++17")
+          message(STATUS "pybind11 selected C++17 flag")
+        elseif(HAS_CPP14_FLAG)
+          set(PYBIND11_CPP_STANDARD "-std=c++14")
+          message(STATUS "pybind11 selected C++14 flag")
+        elseif(HAS_CPP11_FLAG)
+          set(PYBIND11_CPP_STANDARD "-std=c++11")
+          message(STATUS "pybind11 selected C++11 flag")
+        else()
+          message(FATAL_ERROR "Unsupported compiler -- pybind11 requires C++11 support!")
+        endif()
+      endif()
+    endif()
 
-    if (HAS_CPP14_FLAG)
-      set(PYBIND11_CPP_STANDARD -std=c++14)
+  # Auto add if CMake >= 3.8 and CXX_FEATURES is not set
+  elseif(NOT PYBIND11_CXX_FEATURES)
+    # IN_LIST introduced in CMAKE 3.3
+    # Safe because this will only activate if CMake >= 3.8
+    cmake_policy(SET CMP0057 NEW)
+
+    # The following only print if running the first time,
+    # and no C++ mode selected
+    if(cxx_std_17 IN_LIST CMAKE_CXX_COMPILE_FEATURES)
+      set(PYBIND11_CXX_FEATURES cxx_std_17)
+      message(STATUS "pybind11 selected C++17 mode, compiler feature")
+    elseif(cxx_std_14 IN_LIST CMAKE_CXX_COMPILE_FEATURES)
+      message(STATUS "pybind11 selected C++14 mode, compiler feature")
+      set(PYBIND11_CXX_FEATURES cxx_std_14)
+    elseif(cxx_std_11 IN_LIST CMAKE_CXX_COMPILE_FEATURES)
+      message(STATUS "pybind11 selected C++11 mode, compiler feature")
+      set(PYBIND11_CXX_FEATURES cxx_std_11)
     else()
-      check_cxx_compiler_flag("-std=c++11" HAS_CPP11_FLAG)
-      if (HAS_CPP11_FLAG)
-        set(PYBIND11_CPP_STANDARD -std=c++11)
+      message(FATAL_ERROR "Unsupported compiler -- pybind11 requires C++11 support!")
+    endif()
+  endif()
+endif()
+
+# Allow the user to override by setting the cache value after the auto-discovery
+# Empty values get filled on first run (PYBIND11_CPP_STANDARD for CMake < 3.8,
+# otherwise PYBIND11_CXX_FEATURES), and then at this point are promoted to the CACHE.
+
+set(PYBIND11_CPP_STANDARD ${PYBIND11_CPP_STANDARD} CACHE STRING
+  "C++ standard flag, e.g. -std=c++11, -std=c++14, /std:c++14.  Defaults to highest supported for CMake 2.8-3.7")
```

Comment:
What happens if nothing is specified and we are on CMake 3.8? Of course it would be good if the highest standard is used even there.